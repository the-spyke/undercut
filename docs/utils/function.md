---
title: Functions
---

- `getDoneItem(value) => IteratorResult` -- returns an IteratorResult with `done` set to `true`.
- `identity(value) => value` -- returns the same `value` the was provided as the first argument.
- `negate(value) => boolean` -- inverts the sign of the numeric `value`.
- `noop() => void` -- an empty function.
- `not(value) => number` -- performs boolean not `!` operation with the vale.
- `rethrow(error) => never` -- immidietly throws provided `error`.
