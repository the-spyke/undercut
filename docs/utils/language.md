---
title: Language Utils
---

- `hasOwnProps(value) => boolean` -- returns `true` if the `value` is an object having own properties, `false` otherwise.
- `isArrayBuffer(value) => boolean` -- returns `true` if the `value` is an `ArrayBuffer`, `false` otherwise.
- `isBoolean(value) => boolean` -- returns `true` if the `value` is of the boolean type, `false` otherwise.
- `isCoroutine(value) => boolean` -- returns `true` if the `value` looks like `Coroutine`, `false` otherwise. The only mandatory method is `next`.
- `isDate(value) => boolean` -- returns `true` if the `value` is a `Date` object, `false` otherwise.
- `isDefined(value) => boolean` -- returns `true` if the `value` is exactly not `undefined`, `false` otherwise.
- `isError(value) => boolean` -- returns `true` if the `value` is an `Error`, `false` otherwise.
- `isFalsy(value) => boolean` -- returns `true` if the `value` converted to Boolean is `false`, `false` otherwise.
- `isFunction(value) => boolean` -- returns `true` if the `value` is a function, `false` otherwise.
- `isIterable(value) => boolean` -- returns `true` if the `value` is an object having `Symbol.iterator` method, `false` otherwise.
- `isIterator(value) => boolean` -- returns `true` if the `value` is an object having `next` method, `false` otherwise.
- `isMap(value) => boolean` -- returns `true` if the `value` is an `Map`, `false` otherwise.
- `isNegative(value) => boolean` -- returns `true` if the `value` is a Number less than zero, `false` otherwise.
- `isNegativeOrZero(value) => boolean` -- returns `true` if the `value` is a Number less or equal to zero, `false` otherwise.
- `isNull(value) => boolean` -- returns `true` if the `value` is exactly `null`, `false` otherwise.
- `isNullish(value) => boolean` -- returns `true` if the `value` is exactly `null` or `undefined`, `false` otherwise.
- `isNumber(value) => boolean` -- returns `true` if the `value` is of the Number type, `false` otherwise. `NaN` is a Number in JavaScript.
- `isNumberValue(value) => boolean` -- same as `isNumber`, but returns `false` for `NaN`.
- `isObject(value) => boolean` -- returns `true` if the `value` is an Object, `false` otherwise. `null` is an Object is JavaScript.
- `isObjectValue(value) => boolean` -- same as `isObject`, but returns `false` for `null`.
- `isObserver(value) => boolean` -- returns `true` if the `value` is an object with the `next` method, `false` otherwise.
- `isPlainObject(value) => boolean` -- returns `true` if the `value` is an object having no prototype or an Object.prototype, `false` otherwise.
- `isPositive(value) => boolean` -- returns `true` if the `value` is a Number greater that zero, `false` otherwise.
- `isPositiveOrZero(value) => boolean` -- returns `true` if the `value` is a Number equal or greater than zero, `false` otherwise.
- `isRegExp(value) => boolean` -- returns `true` if the `value` is a `RegExp`, `false` otherwise.
- `isSet(value) => boolean` -- returns `true` if the `value` is a `Set`, `false` otherwise.
- `isString(value) => boolean` -- returns `true` if the `value` is of the String type, `false` otherwise.
- `isSymbol(value) => boolean` -- returns `true` if the `value` is of the Symbol type, `false` otherwise.
- `isTruthy(value) => boolean` -- returns `true` if the `value` converted to Boolean is `true`, `false` otherwise.
- `isUndefined(value) => boolean` -- returns `true` if the `value` is exactly `undefined`, `false` otherwise.
- `isWeakMap(value) => boolean` -- returns `true` if the `value` is a `WeakMap`, `false` otherwise.
- `isWeakSet(value) => boolean` -- returns `true` if the `value` is a `WeakSet`, `false` otherwise.
