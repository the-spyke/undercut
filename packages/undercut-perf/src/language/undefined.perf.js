import { runPerfSuite } from "../perf.js";

const x = [1, 2, 4];

function isUndefinedEqual(value) {
	return value === undefined;
}

function isUndefinedType(value) {
	return typeof value === `undefined`;
}

function isDefined(value) {
	return typeof value !== `undefined`;
}

runPerfSuite(`min : check for undefined`, {
	"equality": () => {
		return isUndefinedEqual(x) === false;
	},
	"typeof": () => {
		return isUndefinedType(x) === false;
	},
	"defined": () => {
		return isDefined(x) === true;
	}
});

function isTruthyNegation(value) {
	return !!value;
}

function isTruthyTenary(value) {
	return value ? true : false;
}

function isTruthyBoolean(value) {
	return Boolean(value);
}

runPerfSuite(`min : check is truthy`, {
	"negation": () => {
		return isTruthyNegation(x) === true;
	},
	"tenary": () => {
		return isTruthyTenary(x) === true;
	},
	"boolean": () => {
		return isTruthyBoolean(x) === true;
	}
});
