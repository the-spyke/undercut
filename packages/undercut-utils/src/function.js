export function getDoneItem(value) {
	return { value, done: true };
}

/**
 * @param {any} [value]
 * @returns {any}
 */
export function identity(value) {
	return value;
}

/**
 * Changes the sign of the `value`.
 * @param {number} value
 * @returns {number}
 */
export function negate(value) {
	return -value;
}

/**
 * @returns {void}
 */
export function noop() {
	// Empty.
}

/**
 * Applies logical not `!` operator to the `value`.
 * @param {any} value
 * @returns {boolean}
 */
export function not(value) {
	return !value;
}

/**
 * @param {any} error
 * @returns {void}
 */
export function rethrow(error) {
	throw error;
}
