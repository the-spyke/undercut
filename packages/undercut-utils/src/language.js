const ObjectPrototype = Object.getPrototypeOf({});

/**
 * Gets value's type from "Object.prototype.toString" method. For example, an async function will return "AsyncFunction".
 * Use for data passed between different contexts, where "instanceof" doesn't work.
 * @param {any} value
 * @returns {string}
 */
export function getObjectType(value) {
	const type = ObjectPrototype.toString.call(value);

	return type.substring(8, type.length - 1); // Remove "[object " and the closing bracket.
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function hasOwnProps(value) {
	return isObjectValue(value) && Object.keys(value).length > 0;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isArrayBuffer(value) {
	return value instanceof ArrayBuffer;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isBoolean(value) {
	return typeof value === `boolean`;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isCoroutine(value) {
	return isObjectValue(value) && isFunction(value.next);
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isDate(value) {
	return value instanceof Date;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isDefined(value) {
	return value !== undefined;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isError(value) {
	return value instanceof Error;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isFalsy(value) {
	return !value;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isFunction(value) {
	return typeof value === `function`;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isIterable(value) {
	return !isNullish(value) && isFunction(value[Symbol.iterator]);
}

export { isCoroutine as isIterator };

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isMap(value) {
	return value instanceof Map;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isNegative(value) {
	return isNumberValue(value) && value < 0;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isNegativeOrZero(value) {
	return isNumberValue(value) && value <= 0;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isNull(value) {
	return value === null;
}

/**
 * Checks if the `value` is `null` or `undefined`.
 * @param {any} value
 * @returns {boolean}
 */
export function isNullish(value) {
	return value == null;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isNumber(value) {
	return typeof value === `number`;
}

/**
 * Checks if the `value` is a number, but not `NaN`.
 * @param {any} value
 * @returns {boolean}
 */
export function isNumberValue(value) {
	return isNumber(value) && !Number.isNaN(value);
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isObject(value) {
	return typeof value === `object`;
}

/**
 * Checks if the `value` is an object, but not `null`.
 * @param {any} value
 * @returns {boolean}
 */
export function isObjectValue(value) {
	return isObject(value) && value !== null;
}

export { isCoroutine as isObserver };

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isPlainObject(value) {
	if (!isObjectValue(value)) {
		return false;
	}

	const proto = Object.getPrototypeOf(value);

	return proto === ObjectPrototype || proto === null;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isPositive(value) {
	return isNumberValue(value) && value > 0;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isPositiveOrZero(value) {
	return isNumberValue(value) && value >= 0;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isRegExp(value) {
	return value instanceof RegExp;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isSet(value) {
	return value instanceof Set;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isString(value) {
	return typeof value === `string`;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isSymbol(value) {
	return typeof value === `symbol`;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isTruthy(value) {
	return Boolean(value);
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isUndefined(value) {
	return value === undefined;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isWeakMap(value) {
	return value instanceof WeakMap;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isWeakSet(value) {
	return value instanceof WeakSet;
}
