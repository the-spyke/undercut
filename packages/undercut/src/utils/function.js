/**
 * @param {any} [value]
 * @returns {any}
 */
export function identity(value) {
	return value;
}

/**
 * Applies boolean negation to the predicate's return value.
 * @param {Function} predicate
 * @returns {Function}
 */
export function negate(predicate) {
	return function (...args) {
		return !predicate(...args);
	};
}

/**
 * Multiplies the predicate's return value by `-1`.
 * @param {Function} predicate
 * @returns {Function}
 */
export function negateSign(operation) {
	return function (...args) {
		return -1 * operation(...args);
	};
}

/**
 * @returns {void}
 */
export function noop() {
	// Empty.
}

/**
 * @param {any} error
 * @returns {void}
 */
export function rethrow(error) {
	throw error;
}
