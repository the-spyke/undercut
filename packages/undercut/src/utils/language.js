/**
 * A re-export of `Array.isArray` for consistency.
 */
export const isArray = Array.isArray;

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isBoolean(value) {
	return typeof value === `boolean`;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isCoroutine(value) {
	return isObjectValue(value) && isFunction(value.next);
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isDefined(value) {
	return typeof value !== `undefined`;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isFalsy(value) {
	return !value;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isFunction(value) {
	return typeof value === `function`;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isIterable(value) {
	return !isNullish(value) && isFunction(value[Symbol.iterator]);
}

export { isCoroutine as isIterator };

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isNegative(value) {
	return isNumberValue(value) && value < 0;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isNegativeOrZero(value) {
	return isNumberValue(value) && value <= 0;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isNull(value) {
	return value === null;
}

/**
 * Checks if the `value` is `null` or `undefined`.
 * @param {any} value
 * @returns {boolean}
 */
export function isNullish(value) {
	return value == null;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isNumber(value) {
	return typeof value === `number`;
}

/**
 * Checks if the `value` is a number, but not `NaN`.
 * @param {any} value
 * @returns {boolean}
 */
export function isNumberValue(value) {
	return isNumber(value) && !Number.isNaN(value);
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isObject(value) {
	return typeof value === `object`;
}

/**
 * Checks if the `value` is an object, but not `null`.
 * @param {any} value
 * @returns {boolean}
 */
export function isObjectValue(value) {
	return isObject(value) && value !== null;
}

export { isCoroutine as isObserver };

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isPositive(value) {
	return isNumberValue(value) && value > 0;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isPositiveOrZero(value) {
	return isNumberValue(value) && value >= 0;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isString(value) {
	return typeof value === `string`;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isSymbol(value) {
	return typeof value === `symbol`;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isTruthy(value) {
	return !!value;
}

/**
 * @param {any} value
 * @returns {boolean}
 */
export function isUndefined(value) {
	return typeof value === `undefined`;
}
