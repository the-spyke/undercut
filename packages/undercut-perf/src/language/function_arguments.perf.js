import { zipWith as zipArray } from "@undercut/pull";
import { simulatePull } from "@undercut/testing";
import { identity } from "@undercut/utils";

import { runPerfSuite } from "../perf.js";

const a = [1, 2, 3, 4, 5];
const b = [6, 7, 8, 9, 0];
const c = [4, 2, 7, 5, 1];
const d = [9, 0, 3, 6, 8];

function arrayDestruct2([x, y]) {
	return x + y;
}

function arrayDestruct3([x, y, z]) {
	return x + y + z;
}

function arrayDestruct4([x, y, z, z2]) {
	return x + y + z + z2;
}

function restArgs(...args) {
	return args;
}

function manual2(x, y) {
	return [x, y];
}

function manual3(x, y, z) {
	return [x, y, z];
}

function manual4(x, y, z, z2) {
	return [x, y, z, z2];
}

runPerfSuite(`zip-reducer-argumemts`, {
	"fixed-x2": () => {
		return simulatePull(zip1(manual2, b), a);
	},
	"fixed-x3": () => {
		return simulatePull(zip3(manual3, b, c), a);
	},
	"fixed-x4": () => {
		return simulatePull(zip4(manual4, b, c, d), a);
	},

	"multiple-array-x2": () => {
		return simulatePull(zipArray(identity, b), a);
	},
	"multiple-array-x3": () => {
		return simulatePull(zipArray(identity, b, c), a);
	},
	"multiple-array-x4": () => {
		return simulatePull(zipArray(identity, b, c, d), a);
	},

	"multiple-array-destruct-x2": () => {
		return simulatePull(zipArray(arrayDestruct2, b), a);
	},
	"multiple-array-destruct-x3": () => {
		return simulatePull(zipArray(arrayDestruct3, b, c), a);
	},
	"multiple-array-destruct-x4": () => {
		return simulatePull(zipArray(arrayDestruct4, b, c, d), a);
	},

	"multiple-args-x2": () => {
		return simulatePull(zipArgs(restArgs, b), a);
	},
	"multiple-args-x3": () => {
		return simulatePull(zipArgs(restArgs, b, c), a);
	},
	"multiple-args-x4": () => {
		return simulatePull(zipArgs(restArgs, b, c, d), a);
	},

	"multiple-args-manual-x2": () => {
		return simulatePull(zipArgs(manual2, b), a);
	},
	"multiple-args-manual-x3": () => {
		return simulatePull(zipArgs(manual3, b, c), a);
	},
	"multiple-args-manual-x4": () => {
		return simulatePull(zipArgs(manual4, b, c, d), a);
	},
});

function zip1(factory, source2) {
	return function* (iterable) {
		const iter1 = iterable[Symbol.iterator]();
		const iter2 = source2[Symbol.iterator]();

		while (true) {
			const item1 = iter1.next();
			const item2 = iter2.next();

			if (item1.done && item2.done) {
				break;
			}

			yield factory(item1.value, item2.value);
		}
	};
}

function zip3(factory, source2, source3) {
	return function* (iterable) {
		const iter1 = iterable[Symbol.iterator]();
		const iter2 = source2[Symbol.iterator]();
		const iter3 = source3[Symbol.iterator]();

		while (true) {
			const item1 = iter1.next();
			const item2 = iter2.next();
			const item3 = iter3.next();

			if (item1.done && item2.done && item3.done) {
				break;
			}

			yield factory(item1.value, item2.value, item3.value);
		}
	};
}

function zip4(factory, source2, source3, source4) {
	return function* (iterable) {
		const iter1 = iterable[Symbol.iterator]();
		const iter2 = source2[Symbol.iterator]();
		const iter3 = source3[Symbol.iterator]();
		const iter4 = source4[Symbol.iterator]();

		while (true) {
			const item1 = iter1.next();
			const item2 = iter2.next();
			const item3 = iter3.next();
			const item4 = iter4.next();

			if (item1.done && item2.done && item3.done && item4.done) {
				break;
			}

			yield factory(item1.value, item2.value, item3.value, item4.value);
		}
	};
}

function zipArgs(factory, ...sources) {
	return function* (iterable) {
		const iters = [iterable[Symbol.iterator]()];

		sources.forEach(source => iters.push(source[Symbol.iterator]()));

		while (true) {
			let done = true;

			const values = iters.map(iter => {
				const item = iter.next();

				done = done && item.done;

				return item.value;
			});

			if (done) {
				break;
			}

			yield factory(...values);
		}
	};
}
