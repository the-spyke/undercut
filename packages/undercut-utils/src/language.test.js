import { expect, test } from "@jest/globals";

import { noop } from "./function.js";

import * as language from "./language.js";

function getTestValues() {
	/* eslint-disable brace-style */
	return [
		[10, undefined],
		[11, false],
		[12, true],
		[13, -123, `Negative integer`],
		[14, 0, `Zero`],
		[15, 123, `Positive integer`],
		[16, ``, `Empty string`],
		[17, `test`, `String`],
		[18, Symbol(), `Symbol`],
		[19, null],
		[20, {}, `Plain empty object`],
		[21, { x: 0 }, `Plain object with props`],
		[22, [], `Empty array`],
		[23, [1, 2, 3], `Array with values`],
		[24, noop, `Function`],
		[25, /test/, `RegExp`],
		[26, new Map(), `Map`],
		[27, new Set(), `Set`],
		[28, new WeakMap(), `WeakMap`],
		[29, new WeakSet(), `WeakSet`],
		[30, new ArrayBuffer(), `ArrayBuffer`],
		[31, new Int32Array(), `Int32Array`],
		[32, new Error(`Error message`), `Error`],
		[33, new Date(), `Date`],
		[34, function* () { yield 1; }, `Generator function`],
		[35, (function* () { yield 1; })(), `Generator object`],
		[36, {
			next() { return { value: 36, done: false }; }
		}, `Manual coroutine`],
		[37, {
			[Symbol.iterator]() {
				return {
					next() { return { value: 37, done: false }; }
				};
			}
		}, `Manual iterable`],
		[38, {
			next() { return { value: 38, done: false }; },
			close() { /* Empty. */ },
		}, `Manual closable coroutine`],
		[39, {
			next() { return { value: 39, done: false }; },
			throw(e) { throw e; },
		}, `Manual throwable coroutine`],
		[40, {
			next() { return { value: 40, done: false }; },
			close() { /* Empty. */ },
			throw(e) { throw e; },
		}, `Manual full coroutine`],
		[41, NaN, `NaN`],
		[42, Infinity, `+Infinity`],
		[43, -Infinity, `-Infinity`],
		[44, class User { }, `User class`],
		[45, getInstance(class UserEmpty { }), `EmptyUser object`],
		[46, getInstance(class UserHello { hello() { return `hello`; } }), `UserHello object with hello() method`],
		[47, -123.321, `Negative fractional number`],
		[48, -0, `Negative zero`],
		[49, 123.321, `Positive fractional number`],
		[50, async function () { await Promise.resolve(); }, `Async function`],
	];
	/* eslint-enable brace-style */
}

test(`getObjectType`, () => {
	const types = getTestValues().map(tv => language.getObjectType(tv[1]));

	expect(types).toMatchSnapshot();
});

test(`hasOwnProps`, () => {
	expect(positiveResults(language.hasOwnProps)).toEqual(resultsByIndex([
		21, 23, 36, 38, 39, 40,
	]));
});

test(`isArrayBuffer`, () => {
	expect(positiveResults(language.isArrayBuffer)).toEqual(resultsByIndex([
		30,
	]));
});

test(`isBoolean`, () => {
	expect(positiveResults(language.isBoolean)).toEqual(resultsByIndex([
		11, 12,
	]));
});

test(`isCoroutine`, () => {
	expect(positiveResults(language.isCoroutine)).toEqual(resultsByIndex([
		35, 36, 38, 39, 40,
	]));
});

test(`isDate`, () => {
	expect(positiveResults(language.isDate)).toEqual(resultsByIndex([
		33,
	]));
});

test(`isDefined`, () => {
	expect(positiveResults(language.isDefined)).toEqual(resultsByExcluded([
		10,
	]));
});

test(`isError`, () => {
	expect(positiveResults(language.isError)).toEqual(resultsByIndex([
		32,
	]));
});

test(`isFalsy`, () => {
	expect(positiveResults(language.isFalsy)).toEqual(resultsByIndex([
		10, 11, 14, 16, 19, 41, 48,
	]));
});

test(`isFunction`, () => {
	expect(positiveResults(language.isFunction)).toEqual(resultsByIndex([
		24, 34, 44, 50,
	]));
});

test(`isIterable`, () => {
	expect(positiveResults(language.isIterable)).toEqual(resultsByIndex([
		16, 17, 22, 23, 26, 27, 31, 35, 37,
	]));
});

test(`isIterator`, () => {
	expect(positiveResults(language.isIterator)).toEqual(resultsByIndex([
		35, 36, 38, 39, 40,
	]));
});

test(`isMap`, () => {
	expect(positiveResults(language.isMap)).toEqual(resultsByIndex([
		26,
	]));
});

test(`isNegative`, () => {
	expect(positiveResults(language.isNegative)).toEqual(resultsByIndex([
		13, 43, 47,
	]));
});

test(`isNegativeOrZero`, () => {
	expect(positiveResults(language.isNegativeOrZero)).toEqual(resultsByIndex([
		13, 14, 43, 47, 48,
	]));
});

test(`isNull`, () => {
	expect(positiveResults(language.isNull)).toEqual(resultsByIndex([
		19,
	]));
});

test(`isNullish`, () => {
	expect(positiveResults(language.isNullish)).toEqual(resultsByIndex([
		10, 19,
	]));
});

test(`isNumber`, () => {
	expect(positiveResults(language.isNumber)).toEqual(resultsByIndex([
		13, 14, 15, 41, 42, 43, 47, 48, 49,
	]));
});

test(`isNumberValue`, () => {
	expect(positiveResults(language.isNumberValue)).toEqual(resultsByIndex([
		13, 14, 15, 42, 43, 47, 48, 49,
	]));
});

test(`isObject`, () => {
	expect(positiveResults(language.isObject)).toEqual(resultsByExcluded([
		10, 11, 12, 13, 14, 15, 16, 17, 18, 24, 34, 41, 42, 43, 44, 47, 48, 49, 50,
	]));
});

test(`isObjectValue`, () => {
	expect(positiveResults(language.isObjectValue)).toEqual(resultsByExcluded([
		10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 24, 34, 41, 42, 43, 44, 47, 48, 49, 50,
	]));
});

test(`isObserver`, () => {
	expect(positiveResults(language.isObserver)).toEqual(resultsByIndex([
		35, 36, 38, 39, 40,
	]));
});

test(`isPlainObject`, () => {
	expect(positiveResults(language.isPlainObject)).toEqual(resultsByIndex([
		20, 21, 36, 37, 38, 39, 40,
	]));
});

test(`isPositive`, () => {
	expect(positiveResults(language.isPositive)).toEqual(resultsByIndex([
		15, 42, 49,
	]));
});

test(`isPositiveOrZero`, () => {
	expect(positiveResults(language.isPositiveOrZero)).toEqual(resultsByIndex([
		14, 15, 42, 48, 49,
	]));
});

test(`isRegExp`, () => {
	expect(positiveResults(language.isRegExp)).toEqual(resultsByIndex([
		25,
	]));
});

test(`isSet`, () => {
	expect(positiveResults(language.isSet)).toEqual(resultsByIndex([
		27,
	]));
});

test(`isString`, () => {
	expect(positiveResults(language.isString)).toEqual(resultsByIndex([
		16, 17,
	]));
});

test(`isSymbol`, () => {
	expect(positiveResults(language.isSymbol)).toEqual(resultsByIndex([
		18,
	]));
});

test(`isTruthy`, () => {
	expect(positiveResults(language.isTruthy)).toEqual(resultsByExcluded([
		10, 11, 14, 16, 19, 41, 48,
	]));
});

test(`isUndefined`, () => {
	expect(positiveResults(language.isUndefined)).toEqual(resultsByIndex([
		10,
	]));
});

test(`isWeakMap`, () => {
	expect(positiveResults(language.isWeakMap)).toEqual(resultsByIndex([
		28,
	]));
});

test(`isWeakSet`, () => {
	expect(positiveResults(language.isWeakSet)).toEqual(resultsByIndex([
		29,
	]));
});

test(`getTestValues`, () => {
	expect(getTestValues().every((entry, index) => entry[0] - 10 === index)).toBe(true);
});

function getInstance(Class) {
	return new Class();
}

function mapValues([index, value, name]) {
	return `${index} : ${name !== undefined ? name : String(value)}`;
}

function positiveResults(predicate) {
	return getTestValues().filter(e => predicate(e[1])).map(mapValues);
}

function resultsByIndex(indexes) {
	const values = getTestValues();

	return indexes.map(i => values[i - 10]).map(mapValues);
}

function resultsByExcluded(indexes) {
	const exclude = new Set(indexes);

	return getTestValues().filter(e => !exclude.has(e[0])).map(mapValues);
}
