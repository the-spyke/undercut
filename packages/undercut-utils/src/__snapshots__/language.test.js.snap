// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getObjectType 1`] = `
Array [
  "Undefined",
  "Boolean",
  "Boolean",
  "Number",
  "Number",
  "Number",
  "String",
  "String",
  "Symbol",
  "Null",
  "Object",
  "Object",
  "Array",
  "Array",
  "Function",
  "RegExp",
  "Map",
  "Set",
  "WeakMap",
  "WeakSet",
  "ArrayBuffer",
  "Int32Array",
  "Error",
  "Date",
  "GeneratorFunction",
  "Generator",
  "Object",
  "Object",
  "Object",
  "Object",
  "Object",
  "Number",
  "Number",
  "Number",
  "Function",
  "Object",
  "Object",
  "Number",
  "Number",
  "Number",
  "AsyncFunction",
]
`;
